### ISSUE TITLE
Improve piracy attraction calculation to accomodate for end-effects.

### ISSUE THREAD

**Issue by MasterOfGrey** [endless-sky/endless-sky#5192]
**Is your feature request related to a problem? Please describe.**
The pirate attraction/deterrence is intended to act as a counter to endless snowballing of trade fleets. However, early game its very easy to trip over it, and late game carriers with cargo fighters/drones can trip over it easily because each cargo parasite adds ~1.8-2 attractiveness points and 0 deterrence points, but the deterrence of the parent ship just doesn't outweight it.
Essentially, early game it's too easy to trip over it with 3 star barges, and late game cargo on fighters/drones is disproportionately weighted compared to regular cargo in ships.

**Describe the solution you'd like**
As far as I can tell this is the section that needs modifying (line 1000 onwards in PlayerInfo.cpp - shown below) but I am not super code savvy, nor fully versed on how to address this. I think though that something as simple as making the -1.8 into a -1.9 or -2.0 could be workable. 

Alternatively, a two part solution where there is:
- some kind of flat reduction if the category is Fighter or Drone might be good, since they are protected by their carrier, they're not as vulnerable proportionally as this formula shows them to be.
- consideration of smaller cargo spaces counting less, the current setup gives a progressively smaller impact on attravctivness for incrementally larger cargo holds, but this means small cargo ships are over-represented. (the -2.0 could somewhat cover this but you could use some kind of ratio with the sqaure root of ship strength because smaller ships would count as less attractive than lumbering big ships then which would reflect the lower reward for pirates and the greater ease with which a small ship can escape.)

pair<double, double> PlayerInfo::RaidFleetFactors() const
{
	double attraction = 0.;
	double deterrence = 0.;
	for(const shared_ptr<Ship> &ship : Ships())
	{
		if(ship->IsParked() || ship->IsDestroyed())
			continue;
		
		attraction += .4 * sqrt(ship->Attributes().Get("cargo space")) - 1.8;
		for(const Hardpoint &hardpoint : ship->Weapons())
			if(hardpoint.GetOutfit())
			{
				const Outfit *weapon = hardpoint.GetOutfit();
				if(weapon->Ammo() && !ship->OutfitCount(weapon->Ammo()))
					continue;
				double damage = weapon->ShieldDamage() + weapon->HullDamage();
				deterrence += .12 * damage / weapon->Reload();
			}
	}
	
	return make_pair(attraction, deterrence);
}

	double attractionLevel = max(0., log2(max(factors.first, 0.)));
	double deterrenceLevel = max(0., log2(max(factors.second, 0.)));

**Comment by Amazinite** [comment]
Given the discussions after https://github.com/endless-sky/endless-sky/commit/cbb7c6f5f5a6c566e39795590b2a1b638d83c65e, I was thinking of perhaps having your number of ships factor into this equation somehow. Since the intended purpose of the raid system is to deter against freighter snowballing, not having raid fleets show up if your fleet is below a certain size (excluding fighters) might make sense.

**Comment by MasterOfGrey** [comment]
> Since the intended purpose of the raid system is to deter against freighter snowballing, not having raid fleets show up if your fleet is below a certain size (excluding fighters) might make sense.

That's definitely a possible solution that fixes the early game problem, though I'm not sure how well it would cover strong carrier fleets having their cargo attractiveness misrepresented.

**Comment by InfiniteDonuts** [comment]
The pirate attacks from this game mechanic are ridiculous. I had a fleet of 3 Freighters, and pirates were jumping into the _paradise worlds_ to kill me. The PARADISE WORLDS! Where the Navy is strongest! Maybe add a bit of code to increase/decrease the probability of pirate attacks depending on where you are, and how close you are to a pirate-controlled system?

**Comment by Zitchas** [comment]
Yeah, there might be grounds for some reduction of piracy there...

That being said, EVE-online has a good counter-example of players raiding supposedly safe "high-sec" (high-security) space where highly-lethal and fast spawning  NPC fleets are quick to pounce on any attackers. Pirate raiding into secure territory does happen, even under threat of insta-death. It just happens less.

**Comment by Terin** [comment]
It definitely makes sense that, if there's a good score to be had, you'd have pirate fleets diving deep into hostile territory for a chance at that. However, that should be triggered by like, a half-dozen bulk freighters that have zero defenses, and not a few star barges.

**Comment by InfiniteDonuts** [comment]
> It definitely makes sense that, if there's a good score to be had, you'd have pirate fleets diving deep into hostile territory for a chance at that. However, that should be triggered by like, a half-dozen bulk freighters that have zero defenses, and not a few star barges.

Exactly.
One pirate Leviathan jumped into the Castor system and got focus-fired by the six Navy Cruisers that were waiting for it.
Possibly make pirates less suicidal?

**Comment by Hurleveur** [comment]
The recent PR added a way to make early game and late game pirate fleets, whilstaking the system you are in relevant, so truly we're almost there.
I optimized the way there are calculated to only happen when needed, meaning we can use advanced formulas, should we need to.

That said I'd just use cost, honestly, instead of bothering to add everything together.
Indeed, cost is used for assessing the strength of ships, currently. I'd suggest to use deterrence instead, except it would only look at the cost of the weapons (instead of the whole ship), considering AM although somewhat less, and also taking combat abilities (like shields, regen) into account, somehow.
